Programming_Engine/Compute/PopulateTypes.cs

```diff
@@ -0,0 +1,163 @@
+            if (node.Outputs.Count == 1)
+                node.Outputs[0].DataType = node.Method.ReturnType;
+            else if (node.Outputs.Count > 1 && node.Method.ReturnType is oM.Reflection.Interface.IOutput)
+            {
+                Type[] types = node.Method.ReturnType.GetGenericArguments();
+                if (types.Length == node.Outputs.Count)
+                {
+                    for (int i = 0; i < types.Length; i++)
+                        node.Outputs[i].DataType = types[i];
+                }
+            }
+
+            return node;
+        }
+
+        /***************************************************/
+
+        public static ConstructorNode PopulateTypes(this ConstructorNode node)
+        {
+            if (node == null || node.Constructor == null)
+                return node;
+
+            ParameterInfo[] methodParams = node.Constructor.GetParameters();
+            if (node.Inputs.Count != methodParams.Length)
+                return node;
+
+            for (int i = 0; i < methodParams.Length; i++)
+            {
+                node.Inputs[i].DataType = methodParams[i].ParameterType;
+                if (methodParams[i].HasDefaultValue)
+                    node.Inputs[i].DefaultValue = methodParams[i].DefaultValue;
+            }
+
+
+            if (node.Outputs.Count > 0)
+                node.Outputs[0].DataType = node.Constructor.DeclaringType;
+
+            return node;
+        }
+
+        /***************************************************/
+
+        public static InitialiserNode PopulateTypes(this InitialiserNode node)
+        {
+            if (node == null || node.ObjectType == null)
+                return node;
+
+            PropertyInfo[] properties = node.ObjectType.GetProperties();
+            object instance = Activator.CreateInstance(node.ObjectType);
+
+            foreach (ReceiverParam input in node.Inputs)
+            {
+                PropertyInfo match = properties.FirstOrDefault(x => x.Name == input.Name);
+                if (match != null)
+                {
+                    input.DataType = match.PropertyType;
+                    input.DefaultValue = match.GetValue(instance);
+                }
+            }
+
+            if (node.Outputs.Count > 0)
+                node.Outputs[0].DataType = node.ObjectType;
+
+            return node;
+        }
+
+        /***************************************************/
+
+        public static TypeNode PopulateTypes(this TypeNode node)
+        {
+            if (node == null || node.Type == null)
+                return node;
+
+            foreach (DataParam param in node.Outputs)
+                param.DataType = typeof(Type);
+
+            return node;
+        }
+
+        /***************************************************/
+        /**** Private Methods                           ****/
+        /***************************************************/
+
+        public static INode PopulateTypes(this INode node)
```

Comment:
If this is meant to be public then Description, input, and output attributes needed here.

However, it is under the heading "Private Methods" - so I guess I'm querying whether it should be private instead? If it is private then obviously the above attributes aren't needed